File: backend/src/main/java/lan/dk/podcastserver/business/TagBusiness.java
package lan.dk.podcastserver.business;

import io.vavr.collection.List;
import io.vavr.collection.Set;
import lan.dk.podcastserver.entity.Tag;
import lan.dk.podcastserver.repository.TagRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Component;

import java.util.UUID;

/**
 * Created by kevin on 07/06/2014.
 */
@Slf4j
@Component
@RequiredArgsConstructor
public class TagBusiness {

    private final TagRepository tagRepository;

    public List<Tag> findAll() {
        return List.ofAll(tagRepository.findAll());
    }

    public Tag findOne(UUID id) {
        return tagRepository.findById(id).orElseThrow(() -> new Error("Tag with ID "+ id +" not found"));
    }

    public Set<Tag> findByNameLike(String name) {
        return tagRepository.findByNameContainsIgnoreCase(name);
    }

    Set<Tag> getTagListByName(Set<Tag> tagList) {
        return tagList.map(t -> findByName(t.getName()));
    }

    private Tag findByName(String name) {
        return tagRepository
                .findByNameIgnoreCase(name)
                .getOrElse(() -> tagRepository.save(Tag.builder().name(name).build()));
    }

    public Set<Tag> findAllByName(Set<String> names){
        return names.flatMap(tagRepository::findByNameIgnoreCase);
    }
}

File: backend/src/main/kotlin/com/github/davinkevin/podcastserver/business/TagBusiness.kt
package com.github.davinkevin.podcastserver.business

import com.github.davinkevin.podcastserver.entity.Tag
import lan.dk.podcastserver.repository.TagRepository
import org.springframework.stereotype.Component

/**
 * Created by kevin on 07/06/2014.
 */
@Component
class TagBusiness(val tagRepository: TagRepository) {

    fun getTagListByName(tagList: Set<Tag>): Set<Tag> =
            tagList
                    .mapNotNull { it.name }
                    .map { findByName(it) }
                    .toSet()

    fun findAllByName(names: io.vavr.collection.Set<String>): io.vavr.collection.Set<Tag> =
            names.flatMap { tagRepository.findByNameIgnoreCase(it) }

    private fun findByName(n: String): Tag =
            tagRepository.findByNameIgnoreCase(n)
                    .getOrElse { tagRepository.save(Tag().apply { name = n }) }

}
